admission, kvserver: snapshot integration for disk bandwidth

aadityasondhi opened the pull request:
This patch integrates raft snapshot ingestion with the disk write
mechanism in admission control. The following internal machinery changes
were made to make that possible:
- `SnapshotQueue` was added as an implementation of the `requester`
  interface. Internally, it is a simple FIFO queue unlike the other work
  queue, since we can make the assumption that all snapshots are of the
  same priority and are processed as system tenant requests.
- A new `kvStoreTokenChildGranter` was created to grant tokens to
  snapshot requests.
- We now have a `StoreWorkType` that differentiates `regular`,
  `elastic`, and `snapshot` work for the store granters. This was
  necessary because snapshots do not incur the same write-amp as the
  other work types – they land into L6 of the LSM due to excises. We
  also only want these requests to be subject to pacing based on disk
  bandwidth.
- We now prioritize store writes in the following order: `regular`,
  `snapshot`, `elastic`.
- The `demuxHandle` of the `GrantCoordinator` now uses `StoreWorkType`.

The integration point for the `SnapshotQueue` is in `Receive()` where we
use a pacing mechanism to process incoming snapshots. Snapshots are
subject to `snapshotBurstSize` amounts of disk writes before asking for
further admission of the same size. The `multiSSTWriter` uses Pebble's
SST size estimates to maintain a running count of disk writes incurred
by the snapshot ingest. Once the SST is finalized we deduct/return
further tokens.

Closes https://github.com/cockroachdb/cockroach/issues/120708.

Release note (ops change): Admission Control now has an integration for
pacing snapshot ingest traffic based on disk bandwidth.
`kvadmission.store.snapshot_ingest_bandwidth_control.enabled` is used to
turn on this integration. Note that it requires provisioned bandwidth to
be set for the store (or cluster through the cluster setting) for it to
take effect.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/131243)
<!-- Reviewable:end -->

sumeerbhola reviewed (changes requested):
> Reviewed 1 of 6 files at r1, 4 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131243)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi)

___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 706 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O7fBjN379IEKloMG3wB:-O7fBjN379IEKloMG3wC:bafwl59) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e779e26b99eaf3f1fec8857abde551b5a8f529/pkg/kv/kvserver/replica_raftstorage.go#L706)):*
> ```Go
> 	// TODO(aaditya): Maybe we should remove this?
> 	if r.store.cfg.KVAdmissionController != nil {
> 		r.store.cfg.KVAdmissionController.SnapshotIngestedOrWritten(
> ```

I think we only use this for fitting the models (you should double check), in which case we should document that and keep it. We don't want snapshot writes to change the models we use for normal operations.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 449 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O7fFGKG2K-bExddiOeZ:-O7fFGKG2K-bExddiOe_:bl9zimp) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e779e26b99eaf3f1fec8857abde551b5a8f529/pkg/kv/kvserver/store_snapshot.go#L449)):*
> ```Go
> 		return err
> 	}
> 	prevWriteBytes := msstw.currSST.DataSize
> ```

I don't think we want to work with `DataSize`. There is a `RawRowWriter.EstimatedSize()` which compactions use to decide when to finish the current sstable. We may need to expose that here if not already accessible. And use that since it estimates the on disk size. It may not be monotonic, in that after we finish a sstable, we should correct to the actual written, which may be less. We could ignore any left-over non-monotonicity at the end of the snapshot, or return tokens.
___
*[`pkg/util/admission/granter.go` line 420 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O7fMhmy9xjSW1IivyOP:-O7fMhmy9xjSW1IivyOQ:b-yz4tii) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e779e26b99eaf3f1fec8857abde551b5a8f529/pkg/util/admission/granter.go#L420)):*
> ```Go
> 	// infrequently (every 15s), and the disk bandwidth limiter is designed to
> 	// generally under admit and only pace elastic work.
> 	adjustedDiskWriteTokens := sg.writeAmpLM.applyLinearModel(count)
> ```

we should add a comment that `adjustedDiskWriteTokens` is not used for `SnapshotIngestStoreWorkType` and shouldn't even compute it in that case.
___
*[`pkg/util/admission/granter.go` line 463 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O7fMKdm5TQuMur5LHK0:-O7fMKdm5TQuMur5LHK1:bmabac2) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e779e26b99eaf3f1fec8857abde551b5a8f529/pkg/util/admission/granter.go#L463)):*
> ```Go
> 		sg.subtractTokensLocked(-count, -count, false)
> 	}
> 	if wc == admissionpb.ElasticWorkClass {
> ```

this should only be done when wt == ElasticStoreWorkType.

Try structuring the logic in this function as a switch statement to see if that makes it simpler.
___
*[`pkg/util/admission/granter.go` line 467 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O7fN8DT-Nofz_jW5cn1:-O7fN8DT-Nofz_jW5cn2:bqu39yb) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e779e26b99eaf3f1fec8857abde551b5a8f529/pkg/util/admission/granter.go#L467)):*
> ```Go
> 	}
> 	// Return tokens to disk bandwidth bucket.
> 	diskTokenCount := sg.writeAmpLM.applyLinearModel(count)
> ```

The linear model should not apply to the snapshot case.

(note to self -- stopped here)
___
*[`pkg/util/admission/snapshot_queue.go` line 34 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O7fLXE_6Pd-ASw2lHPF:-O7fLXE_6Pd-ASw2lHPG:bdb4hte) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e779e26b99eaf3f1fec8857abde551b5a8f529/pkg/util/admission/snapshot_queue.go#L34)):*
> ```Go
> 	mu              struct {
> 		syncutil.Mutex
> 		q []*snapshotWorkItem
> ```

Consider using the queue in pkg/util/queue, since it reduces the number of slice allocations.
___
*[`pkg/util/admission/snapshot_queue.go` line 40 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O7fI2ip8j_aRpISHoXt:-O7fI2ip8j_aRpISHoXu:b-dp77kw) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e779e26b99eaf3f1fec8857abde551b5a8f529/pkg/util/admission/snapshot_queue.go#L40)):*
> ```Go
> func MakeSnapshotQueue(snapshotGranter granter) *SnapshotQueue {
> 	q := &SnapshotQueue{snapshotGranter: snapshotGranter}
> 	q.mu.q = []*snapshotWorkItem{}
> ```

By default this is a nil slice, and this is staying nil, so this is unnecessary.
___
*[`pkg/util/admission/snapshot_queue.go` line 73 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O7fKjx_3WJEH7xDGbO7:-O7fKjx_3WJEH7xDGbO8:bjrchdh) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e779e26b99eaf3f1fec8857abde551b5a8f529/pkg/util/admission/snapshot_queue.go#L73)):*
> ```Go
> 	// We were unable to get tokens for admission, so we queue.
> 	s.mu.Lock()
> 	item := &snapshotWorkItem{
> ```

consider adding a TODO to use a sync.Pool for this, similar to `newWaitingWork` and `releaseWaitingWork`.
___
*[`pkg/util/admission/snapshot_queue.go` line 85 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O7fKxd44XLOeFkCxTxH:-O7fKxd44XLOeFkCxTxI:b-ao38z5) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e779e26b99eaf3f1fec8857abde551b5a8f529/pkg/util/admission/snapshot_queue.go#L85)):*
> ```Go
> 	case <-ctx.Done():
> 		waitDur := time.Now().Sub(item.enqueueingTime)
> 		s.snapshotGranter.returnGrant(item.count)
> ```

We should return the grant only if there was a race and `granted` managed to pop from the queue before context cancellation. Otherwise nothing was granted, so nothing needs to be returned.
___
*[`pkg/util/admission/work_queue.go` line 2152 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O7fHmo38cHEAjgskHMR:-O7fHmo38cHEAjgskHMS:bamskdw) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e779e26b99eaf3f1fec8857abde551b5a8f529/pkg/util/admission/work_queue.go#L2152)):*
> ```Go
> // done, invoke AdmittedWorkDone with the now-known size.
> func (q *StoreWorkQueue) AdmittedWorkDone(h StoreWorkHandle, doneInfo StoreWorkDoneInfo) error {
> 	// TODO(aaditya): Update this or make equivalent for snapshots.
> ```

why is this method relevant for snapshots?


<!-- Sent from Reviewable.io -->

aadityasondhi committed:
admission, kvserver: snapshot integration for disk bandwidth

This patch integrates raft snapshot ingestion with the disk write
mechanism in admission control. The following internal machinery changes
were made to make that possible:
- `SnapshotQueue` was added as an implementation of the `requester`
  interface. Internally, it is a simple FIFO queue unlike the other work
  queue, since we can make the assumption that all snapshots are of the
  same priority and are processed as system tenant requests.
- A new `kvStoreTokenChildGranter` was created to grant tokens to
  snapshot requests.
- We now have a `StoreWorkType` that differentiates `regular`,
  `elastic`, and `snapshot` work for the store granters. This was
  necessary because snapshots do not incur the same write-amp as the
  other work types – they land into L6 of the LSM due to excises. We
  also only want these requests to be subject to pacing based on disk
  bandwidth.
- We now prioritize store writes in the following order: `regular`,
  `snapshot`, `elastic`.
- The `demuxHandle` of the `GrantCoordinator` now uses `StoreWorkType`.

The integration point for the `SnapshotQueue` is in `Receive()` where we
use a pacing mechanism to process incoming snapshots. Snapshots are
subject to `snapshotBurstSize` amounts of disk writes before asking for
further admission of the same size. The `multiSSTWriter` uses Pebble's
SST size estimates to maintain a running count of disk writes incurred
by the snapshot ingest. Once the SST is finalized we deduct/return
further tokens.

Closes #120708.

Release note (ops change): Admission Control now has an integration for
pacing snapshot ingest traffic based on disk bandwidth.
`kvadmission.store.snapshot_ingest_bandwidth_control.enabled` is used to
turn on this integration. Note that it requires provisioned bandwidth to
be set for the store (or cluster through the cluster setting) for it to
take effect.

aadityasondhi reviewed (commented):
TFTR! Addressed all comments, still have a couple things to do:
- unit tests
- hook up the wait metric into the metrics system

But it might be worth a quick review while I do those things in the background.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-:-O7oNQGV4j3yTh1R-vz7:bmy246l)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 706 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O7fBjN379IEKloMG3wB:-O7oMr4CA9NPcugBb9P0:b-uia9j0) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e779e26b99eaf3f1fec8857abde551b5a8f529/pkg/kv/kvserver/replica_raftstorage.go#L706)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I think we only use this for fitting the models (you should double check), in which case we should document that and keep it. We don't want snapshot writes to change the models we use for normal operations.
</blockquote></details>

Done. Yes checked, you are correct. Removed the comment.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 449 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O7fFGKG2K-bExddiOeZ:-O7oMucMCi8R1TkJgoGO:bjfekfk) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e779e26b99eaf3f1fec8857abde551b5a8f529/pkg/kv/kvserver/store_snapshot.go#L449)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't think we want to work with `DataSize`. There is a `RawRowWriter.EstimatedSize()` which compactions use to decide when to finish the current sstable. We may need to expose that here if not already accessible. And use that since it estimates the on disk size. It may not be monotonic, in that after we finish a sstable, we should correct to the actual written, which may be less. We could ignore any left-over non-monotonicity at the end of the snapshot, or return tokens.
</blockquote></details>

Done. Thanks for the pointer, now using `RawRowWriter.EstimatedSize()`.
___
*[`pkg/util/admission/granter.go` line 420 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O7fMhmy9xjSW1IivyOP:-O7oN-GnCSnwDdIcVzt1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e779e26b99eaf3f1fec8857abde551b5a8f529/pkg/util/admission/granter.go#L420)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

we should add a comment that `adjustedDiskWriteTokens` is not used for `SnapshotIngestStoreWorkType` and shouldn't even compute it in that case.
</blockquote></details>

Done.
___
*[`pkg/util/admission/granter.go` line 463 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O7fMKdm5TQuMur5LHK0:-O7oN0Jw3F64vjheQH94:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e779e26b99eaf3f1fec8857abde551b5a8f529/pkg/util/admission/granter.go#L463)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

this should only be done when wt == ElasticStoreWorkType.

Try structuring the logic in this function as a switch statement to see if that makes it simpler.
</blockquote></details>

Done.
___
*[`pkg/util/admission/granter.go` line 467 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O7fN8DT-Nofz_jW5cn1:-O7oN1U04KkRHNm8nEsU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e779e26b99eaf3f1fec8857abde551b5a8f529/pkg/util/admission/granter.go#L467)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

The linear model should not apply to the snapshot case.

(note to self -- stopped here)
</blockquote></details>

Done.
___
*[`pkg/util/admission/snapshot_queue.go` line 34 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O7fLXE_6Pd-ASw2lHPF:-O7oNB26EMjPcAaVAQVE:bbx439j) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e779e26b99eaf3f1fec8857abde551b5a8f529/pkg/util/admission/snapshot_queue.go#L34)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Consider using the queue in pkg/util/queue, since it reduces the number of slice allocations.
</blockquote></details>

Done. This is pretty neat.
___
*[`pkg/util/admission/snapshot_queue.go` line 40 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O7fI2ip8j_aRpISHoXt:-O7oN2eX-VwtWw4AMKbo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e779e26b99eaf3f1fec8857abde551b5a8f529/pkg/util/admission/snapshot_queue.go#L40)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

By default this is a nil slice, and this is staying nil, so this is unnecessary.
</blockquote></details>

Done.
___
*[`pkg/util/admission/snapshot_queue.go` line 73 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O7fKjx_3WJEH7xDGbO7:-O7oN3_s8HcIbFcCn4Wj:b-5l3bzn) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e779e26b99eaf3f1fec8857abde551b5a8f529/pkg/util/admission/snapshot_queue.go#L73)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

consider adding a TODO to use a sync.Pool for this, similar to `newWaitingWork` and `releaseWaitingWork`.
</blockquote></details>

Done. I already had a TODO for it in the file above, but decided to just implement it.
___
*[`pkg/util/admission/snapshot_queue.go` line 85 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O7fKxd44XLOeFkCxTxH:-O7oN94n5XJ6VEctxds3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e779e26b99eaf3f1fec8857abde551b5a8f529/pkg/util/admission/snapshot_queue.go#L85)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

We should return the grant only if there was a race and `granted` managed to pop from the queue before context cancellation. Otherwise nothing was granted, so nothing needs to be returned.
</blockquote></details>

Done.
___
*[`pkg/util/admission/work_queue.go` line 2152 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O7fHmo38cHEAjgskHMR:-O7oNFLCA1U6BRwWEwco:b3cy3dl) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e779e26b99eaf3f1fec8857abde551b5a8f529/pkg/util/admission/work_queue.go#L2152)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

why is this method relevant for snapshots?
</blockquote></details>

Done. Not relevant, I was initially thinking of a different approach to returning tokens. But now we directly do return grant in Admit().


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> Reviewed 3 of 9 files at r3, 9 of 16 files at r4, 2 of 3 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131243)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi)

___
*[`pkg/kv/kvserver/store_snapshot.go` line 921 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8ifcrA8EQWnMcuYEop:-O8ifcrA8EQWnMcuYEoq:bxk7szp) ([raw file](https://github.com/cockroachdb/cockroach/blob/953a045251d857b7b3d35c2d51e6cc392a667524/pkg/kv/kvserver/store_snapshot.go#L921)):*
> ```Go
> 			// Defensive call to account for any discrepancies. The SST sizes should
> 			// have been updated upon closing.
> 			additionalWrites := sstSize - msstw.writeBytes
> ```

What if `additionalWrites` is negative? I am not sure monotonicity is guaranteed in the estimation.
___
*[`pkg/storage/sst_writer.go` line 564 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8ifvOp0MXHStqJsdIG:-O8ifvOqDoJhXt8qCTMr:bsppl73) ([raw file](https://github.com/cockroachdb/cockroach/blob/953a045251d857b7b3d35c2d51e6cc392a667524/pkg/storage/sst_writer.go#L564)):*
> ```Go
> 
> // EstimatedSize returns the underlying RawWriter's estimated size. Note that
> // this size is an estimated as if the writer were to be closed at the time of
> ```

nit: Note that this is an estimate of the size if the writer were ...
___
*[`pkg/util/admission/granter.go` line 441 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8ihFrb6c61WuS2FXr3:-O8ihFrcAkkPGFFAkmau:b-v2m6j3) ([raw file](https://github.com/cockroachdb/cockroach/blob/953a045251d857b7b3d35c2d51e6cc392a667524/pkg/util/admission/granter.go#L441)):*
> ```Go
> 			sg.coordMu.elasticIOTokensUsedByElastic += count
> 			sg.coordMu.diskTokensAvailable.writeByteTokens -= diskWriteTokens
> 			sg.coordMu.diskTokensUsed[admissionpb.ElasticWorkClass].writeByteTokens += diskWriteTokens
> ```

diskTokensUsed are now only used for stats -- they are stashed in `diskBandwidthLimiter` and printed in the ioLoadListener log statement. Consider making this keyed by work type too, so we can differentiate between consumption by snapshots and other elastic work.
___
*[`pkg/util/admission/granter.go` line 475 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8iicYj1exurO76jbJ0:-O8iicYk7k9tz5Bdk6gB:bd8ajs5) ([raw file](https://github.com/cockroachdb/cockroach/blob/953a045251d857b7b3d35c2d51e6cc392a667524/pkg/util/admission/granter.go#L475)):*
> ```Go
> 		sg.coordMu.diskTokensUsed[wc].writeByteTokens -= diskTokenCount
> 	case admissionpb.ElasticStoreWorkType:
> 		sg.coordMu.elasticIOTokensUsedByElastic -= count
> ```

The code repetition seems avoidable. How about lifting this `elasticIOTokensUsedByElastic` adjustment to an if-block before the switch statement. Then this switch statement can merge the regular and elastic cases.
___
*[`pkg/util/admission/granter.go` line 495 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8ijWbiC46QlJauYi3W:-O8ijWbiC46QlJauYi3X:b-9py10o) ([raw file](https://github.com/cockroachdb/cockroach/blob/953a045251d857b7b3d35c2d51e6cc392a667524/pkg/util/admission/granter.go#L495)):*
> ```Go
> 
> // tookWithoutPermissionLocked implements granterWithLockedCalls.
> func (sg *kvStoreTokenGranter) tookWithoutPermissionLocked(count int64, demuxHandle int8) {
> ```

We should merge `tookWithoutPermissionLocked` and `returnGrantLocked`. They do the same thing except with the sign of the count inverted. Otherwise there is a risk of divergence, which seems to be what has happened with this change.
___
*[`pkg/util/admission/granter.go` line 699 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8ikN_E-7L-7oMcD5zN:-O8ikN_E-7L-7oMcD5zO:b-bpr3d8) ([raw file](https://github.com/cockroachdb/cockroach/blob/953a045251d857b7b3d35c2d51e6cc392a667524/pkg/util/admission/granter.go#L699)):*
> ```Go
> 	additionalL0TokensNeeded := actualL0Tokens - originalTokens
> 	sg.subtractTokensLocked(additionalL0TokensNeeded, additionalL0TokensNeeded, false)
> 	if wc == admissionpb.ElasticWorkClass && wt == admissionpb.ElasticStoreWorkType {
> ```

Doesn't ElasticStoreWorkType => ElasticWorkClass?
___
*[`pkg/util/admission/snapshot_queue.go` line 123 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8ilzkPEd3v3a59Z4aO:-O8ilzkQ-x-uVr24JLgK:bmyggvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/953a045251d857b7b3d35c2d51e6cc392a667524/pkg/util/admission/snapshot_queue.go#L123)):*
> ```Go
> 		// Item was cancelled, we can ignore it.
> 		releaseSnapshotWorkItem(item)
> 		return 0
> ```

why don't we loop until there is no cancelled item or the queue is empty?
___
*[`pkg/util/admission/snapshot_queue.go` line 170 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8imow9Df5idT4P2PpW:-O8imow9Df5idT4P2PpX:b-wjqpai) ([raw file](https://github.com/cockroachdb/cockroach/blob/953a045251d857b7b3d35c2d51e6cc392a667524/pkg/util/admission/snapshot_queue.go#L170)):*
> ```Go
> 	case <-ctx.Done():
> 		waitDur := timeutil.Since(item.enqueueingTime).Nanoseconds()
> 		if !item.inQueue {
> ```

we can't read inQueue, or update cancelled (done below) without holding a mutex, since they could be concurrently read/written by the granting.
Also, can you add a comment where `snapshotWorkItem` is declared on which members are not updated after the struct is initialized and which are updated.


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (commented):
TFTR!

I now added a new `SnapshotPacer` to deal with ingesting snapshots and making the interface in `store_snapshot.go` simple. It also abstracts away enough logic so that we can unit test the pacing logic separately. Let me know what you think.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-:-O8rCFkB9Pokr44oVImi:b-2daoe2)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[`pkg/kv/kvserver/store_snapshot.go` line 921 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8ifcrA8EQWnMcuYEop:-O8rB7MI6Ch_t1eCPjQ7:bn64ero) ([raw file](https://github.com/cockroachdb/cockroach/blob/953a045251d857b7b3d35c2d51e6cc392a667524/pkg/kv/kvserver/store_snapshot.go#L921)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

What if `additionalWrites` is negative? I am not sure monotonicity is guaranteed in the estimation.
</blockquote></details>

This is handled by `Admit`. If count is negative, it will return the tokens back to the granter. 
___
*[`pkg/storage/sst_writer.go` line 564 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8ifvOp0MXHStqJsdIG:-O8rBGe6CBd_Pt-Qs7FM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/953a045251d857b7b3d35c2d51e6cc392a667524/pkg/storage/sst_writer.go#L564)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: Note that this is an estimate of the size if the writer were ...
</blockquote></details>

Done.
___
*[`pkg/util/admission/granter.go` line 441 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8ihFrb6c61WuS2FXr3:-O8rBKWg8i2DgB74yzhh:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/953a045251d857b7b3d35c2d51e6cc392a667524/pkg/util/admission/granter.go#L441)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

diskTokensUsed are now only used for stats -- they are stashed in `diskBandwidthLimiter` and printed in the ioLoadListener log statement. Consider making this keyed by work type too, so we can differentiate between consumption by snapshots and other elastic work.
</blockquote></details>

Done. 
___
*[`pkg/util/admission/granter.go` line 475 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8iicYj1exurO76jbJ0:-O8rBZn60brFzc3ucbCe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/953a045251d857b7b3d35c2d51e6cc392a667524/pkg/util/admission/granter.go#L475)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

The code repetition seems avoidable. How about lifting this `elasticIOTokensUsedByElastic` adjustment to an if-block before the switch statement. Then this switch statement can merge the regular and elastic cases.
</blockquote></details>

Done.
___
*[`pkg/util/admission/granter.go` line 495 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8ijWbiC46QlJauYi3W:-O8rBaZWBRASMHCY5Fbg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/953a045251d857b7b3d35c2d51e6cc392a667524/pkg/util/admission/granter.go#L495)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

We should merge `tookWithoutPermissionLocked` and `returnGrantLocked`. They do the same thing except with the sign of the count inverted. Otherwise there is a risk of divergence, which seems to be what has happened with this change.
</blockquote></details>

Done.
___
*[`pkg/util/admission/granter.go` line 699 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8ikN_E-7L-7oMcD5zN:-O8rBboj6qQIQT8bNBvH:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/953a045251d857b7b3d35c2d51e6cc392a667524/pkg/util/admission/granter.go#L699)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Doesn't ElasticStoreWorkType => ElasticWorkClass?
</blockquote></details>

Done. 
___
*[`pkg/util/admission/snapshot_queue.go` line 123 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8ilzkPEd3v3a59Z4aO:-O8rBfncBx8sI-gvOA-x:b-hakueo) ([raw file](https://github.com/cockroachdb/cockroach/blob/953a045251d857b7b3d35c2d51e6cc392a667524/pkg/util/admission/snapshot_queue.go#L123)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

why don't we loop until there is no cancelled item or the queue is empty?
</blockquote></details>

Done. Nice catch!
___
*[`pkg/util/admission/snapshot_queue.go` line 170 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8imow9Df5idT4P2PpW:-O8rBhPI4ETfa4bribOH:b-hakueo) ([raw file](https://github.com/cockroachdb/cockroach/blob/953a045251d857b7b3d35c2d51e6cc392a667524/pkg/util/admission/snapshot_queue.go#L170)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

we can't read inQueue, or update cancelled (done below) without holding a mutex, since they could be concurrently read/written by the granting.
Also, can you add a comment where `snapshotWorkItem` is declared on which members are not updated after the struct is initialized and which are updated.
</blockquote></details>

Done. Nice catch!


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> Reviewed 1 of 16 files at r4, 18 of 22 files at r7, 1 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131243)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi)

___
*[`pkg/util/admission/snapshot_queue.go` line 124 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8w1pWk8n7IT2QYg7_j:-O8w1pWk8n7IT2QYg7_k:bvjlg3v) ([raw file](https://github.com/cockroachdb/cockroach/blob/c38f1e133bdae8094f7612ac6a4d8309ad2bde75/pkg/util/admission/snapshot_queue.go#L124)):*
> ```Go
> 	s.mu.Lock()
> 	defer s.mu.Unlock()
> 	var item *snapshotWorkItem
> ```

The concurrency of `SnapshotQueue` by itself does not need to be high, since there are few concurrent snapshots and the work items are chunky. But `SnapshotQueue` has peer requesters who have high throughput.
So we need to (a) do the least amount of work in `granted`, (b) make the mutex acquisition as narrow as possible in places outside `granted`, so that `granted` doesn't have to wait.

Regarding (a), I couldn't spot anything obvious to remove here. I have a comment elsewhere regarding (b).
___
*[`pkg/util/admission/snapshot_queue.go` line 170 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8w0on9BIsi6x5eKleI:-O8w0on9BIsi6x5eKleJ:bwbuv9u) ([raw file](https://github.com/cockroachdb/cockroach/blob/c38f1e133bdae8094f7612ac6a4d8309ad2bde75/pkg/util/admission/snapshot_queue.go#L170)):*
> ```Go
> 		if shouldRelease {
> 			s.mu.Lock()
> 			defer s.mu.Unlock()
> ```

Locking here seems unnecessary. This code must have sole ownership of the `item` at this point, otherwise putting it back in the `sync.Pool` is a bug.
___
*[`pkg/util/admission/snapshot_queue.go` line 197 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8w3QTm8sI7KjzqTqIY:-O8w3QTn-FWT8wWTs1Uj:b-r8vhfe) ([raw file](https://github.com/cockroachdb/cockroach/blob/c38f1e133bdae8094f7612ac6a4d8309ad2bde75/pkg/util/admission/snapshot_queue.go#L197)):*
> ```Go
> 		// non-ideal behavior, but still provides accurate token accounting.
> 		shouldRelease = false
> 		item.mu.cancelled = true
> ```

Regarding (b) in my comment above, can you lift `item.mu.cancelled=true` above the `shouldRelease = true` and release the mutex.
___
*[`pkg/util/admission/snapshot_queue.go` line 261 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8w40hO3j11XYDaA-xi:-O8w40hO3j11XYDaA-xj:bgvkt7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/c38f1e133bdae8094f7612ac6a4d8309ad2bde75/pkg/util/admission/snapshot_queue.go#L261)):*
> ```Go
> 
> type SnapshotPacer struct {
> 	snapshotQ     snapshotRequester
> ```

Why does `snapshotRequester` need to implement `requester`? `SnapshotPacer` only calls `Admit`.
___
*[`pkg/util/admission/snapshot_queue.go` line 287 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8w4Iq-1_TpBiORd-Wx:-O8w4Iq-1_TpBiORd-Wy:bvwxm6z) ([raw file](https://github.com/cockroachdb/cockroach/blob/c38f1e133bdae8094f7612ac6a4d8309ad2bde75/pkg/util/admission/snapshot_queue.go#L287)):*
> ```Go
> 	if p.intWriteBytes <= SnapshotBurstSize && !final {
> 		return nil
> 	}
> ```

Given `writeBytes` can be negative, it seems possible that `p.intWriteBytes` can become negative. So if `final` happened to be true, we would call `Admit` with a negative value. That is undefined behavior, since `Admit` should only be called with a value > 0.


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (commented):
TFTR! Adressed your comments below. PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-:-O8wNsHQDqdRSf8a-0Sk:bcz3f6a)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[`pkg/util/admission/snapshot_queue.go` line 124 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8w1pWk8n7IT2QYg7_j:-O8wMvDzDZ60R7CTOonB:b336ljm) ([raw file](https://github.com/cockroachdb/cockroach/blob/c38f1e133bdae8094f7612ac6a4d8309ad2bde75/pkg/util/admission/snapshot_queue.go#L124)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

The concurrency of `SnapshotQueue` by itself does not need to be high, since there are few concurrent snapshots and the work items are chunky. But `SnapshotQueue` has peer requesters who have high throughput.
So we need to (a) do the least amount of work in `granted`, (b) make the mutex acquisition as narrow as possible in places outside `granted`, so that `granted` doesn't have to wait.

Regarding (a), I couldn't spot anything obvious to remove here. I have a comment elsewhere regarding (b).
</blockquote></details>

Done. Did (b), I can't think anything to do for (a).
___
*[`pkg/util/admission/snapshot_queue.go` line 170 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8w0on9BIsi6x5eKleI:-O8wN2tX4uCtGBvjwHiX:b-hsk3h6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c38f1e133bdae8094f7612ac6a4d8309ad2bde75/pkg/util/admission/snapshot_queue.go#L170)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Locking here seems unnecessary. This code must have sole ownership of the `item` at this point, otherwise putting it back in the `sync.Pool` is a bug.
</blockquote></details>

Done. Yeah, I initially had no mutex, and naively though the data race detector had a false positive so I added this to silence it. But now I have fixed both, removing this unnecessary mutex acquisition as well as fixing the race condition in the `granted`. 
___
*[`pkg/util/admission/snapshot_queue.go` line 197 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8w3QTm8sI7KjzqTqIY:-O8wNWsYDl6wiohIxg8i:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c38f1e133bdae8094f7612ac6a4d8309ad2bde75/pkg/util/admission/snapshot_queue.go#L197)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Regarding (b) in my comment above, can you lift `item.mu.cancelled=true` above the `shouldRelease = true` and release the mutex.
</blockquote></details>

Done.
___
*[`pkg/util/admission/snapshot_queue.go` line 261 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8w40hO3j11XYDaA-xi:-O8wNO2oBwMBasRWvqpd:b293pjq) ([raw file](https://github.com/cockroachdb/cockroach/blob/c38f1e133bdae8094f7612ac6a4d8309ad2bde75/pkg/util/admission/snapshot_queue.go#L261)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Why does `snapshotRequester` need to implement `requester`? `SnapshotPacer` only calls `Admit`.
</blockquote></details>

Done. You are right. Removed.
___
*[`pkg/util/admission/snapshot_queue.go` line 287 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8w4Iq-1_TpBiORd-Wx:-O8wMHb0ESyC2CYd2Z19:bhrpopq) ([raw file](https://github.com/cockroachdb/cockroach/blob/c38f1e133bdae8094f7612ac6a4d8309ad2bde75/pkg/util/admission/snapshot_queue.go#L287)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Given `writeBytes` can be negative, it seems possible that `p.intWriteBytes` can become negative. So if `final` happened to be true, we would call `Admit` with a negative value. That is undefined behavior, since `Admit` should only be called with a value > 0.
</blockquote></details>

Sent you a message on slack about this, but will also answer here too. Look at my comment around the admit function.
___
*[`pkg/util/admission/snapshot_queue.go` line 159 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8wMYne5aB4TCtFxuFf:-O8wMYne5aB4TCtFxuFg:bte3u2l) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afa15417cca1c5e39ba7f7b9ff44c5e75957424/pkg/util/admission/snapshot_queue.go#L159)):*
> ```Go
> 		return nil
> 	}
> 	if count < 0 {
> ```

About your comment about undefined behavior in Admit. It is actually defined here. I agree that it is different semantics than our other `Admit` functions elsewhere in AC, but it seemed convenient to just do that here. I could lift this into a separate function that wraps `returnGrant`. And the logic of returning tokens could be handled by the pacer in that case. i.e. 
```
if intWrites < 0 { returnGrant(intWrites) }
else { Admit(intWrites) }
```


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 22 files at r7, 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-:-O8wPvwoDzhwYRE1TDLs:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aadityasondhi)

___
*[`pkg/util/admission/snapshot_queue.go` line 159 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/131243#-O8wMYne5aB4TCtFxuFf:-O8wPXWdAOL8WvNH-mgR:b-r1jiny) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afa15417cca1c5e39ba7f7b9ff44c5e75957424/pkg/util/admission/snapshot_queue.go#L159)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

About your comment about undefined behavior in Admit. It is actually defined here. I agree that it is different semantics than our other `Admit` functions elsewhere in AC, but it seemed convenient to just do that here. I could lift this into a separate function that wraps `returnGrant`. And the logic of returning tokens could be handled by the pacer in that case. i.e. 
```
if intWrites < 0 { returnGrant(intWrites) }
else { Admit(intWrites) }
```
</blockquote></details>

What you already have is good -- I'd forgotten about this.


<!-- Sent from Reviewable.io -->

aadityasondhi commented:
bors r=sumeerbhola
